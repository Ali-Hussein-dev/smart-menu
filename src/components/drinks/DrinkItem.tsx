import { useDisclosure } from '@chakra-ui/react'
import { LongText } from '../leaf-node'
import { Price } from '../leaf-node'

type Props = Menu.Item
//=======================
const DrinkItem: React.FC<{
  item: Props
  hasHH?: boolean
  sizeUnit?: Menu.SizeUnit
}> = ({ item, hasHH, sizeUnit }) => {
  // hooks
  //--------------------------------------
  // functions
  const { isOpen, onToggle } = useDisclosure()
  //--------------------------------------
  return (
    <div className="mb-3 text-blueGray-600">
      <div className="flex justify-between">
        <div className="pr-6">
          <span className="font-medium text-blueGray-700">{item.name}</span>
          {item.description && <LongText>{item.description}</LongText>}
          {item.ingredients && <LongText>{item.ingredients}</LongText>}
          {item.brandName && <LongText>{item.brandName}</LongText>}
          {item.filler && <LongText>{item.filler}</LongText>}
        </div>
        <Price
          hasHH={hasHH}
          sizeUnit={sizeUnit}
          isInfo={false}
          price={item.price}
          size={item.size}
          onToggle={onToggle}
          isOpen={isOpen}
        />
      </div>
    </div>
  )
}
export default DrinkItem
