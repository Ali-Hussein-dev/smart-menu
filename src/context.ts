import * as React from 'react'

interface action {
  type:
    | 'onHH'
    | 'offHH'
    | 'onAvailable'
    | 'offAvailable'
    | 'onLunch'
    | 'offLunch'
}
interface CtxType {
  state: Menu.MegaObj
  dispatch?: React.Dispatch<action>
}
export const Ctx = React.createContext<CtxType>(null)

export const reducer = (
  state: Menu.MegaObj,
  action: action
): Menu.MegaObj | null => {
  switch (action.type) {
    case 'onHH':
      return {
        ...state,
        meta: {
          ...state.meta,
          isHH: true,
        },
      }
    case 'offHH':
      return {
        ...state,
        meta: {
          ...state.meta,
          isHH: false,
        },
      }
    case 'onAvailable':
      return {
        ...state,
        meta: {
          ...state.meta,
          isCocktailAvailable: true,
        },
      }
    case 'offAvailable':
      return {
        ...state,
        meta: {
          ...state.meta,
          isCocktailAvailable: false,
        },
      }
    case 'onLunch':
      return {
        ...state,
        meta: {
          ...state.meta,
          isLunch: true,
        },
      }
    case 'offLunch':
      return {
        ...state,
        meta: {
          ...state.meta,
          isLunch: false,
        },
      }
    default:
      throw Error('Unknown action')
  }
}
