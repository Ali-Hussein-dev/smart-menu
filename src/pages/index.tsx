import Head from 'next/head'

export const Home = (): JSX.Element => (
  <div className="grid h-screen place-items-center bg-blueGray-100 text-blueGray-600">
    <Head>
      <title>Smart Gastkarte | HomePage</title>
      {/* <link rel="icon" href="/favicon.ico" /> */}
    </Head>
    <main>
      HomePage
    </main>
  </div>
)

export default Home
