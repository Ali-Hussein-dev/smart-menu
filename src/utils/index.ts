export * from './customHooks'
import dayjs from 'dayjs'

export const validateTime = (time: number): Error | void => {
  if (time > 1440 || time < 0) {
    throw new Error('Time must be between 0-1440')
  }
}
//--------------------------------------1
export const withinTimeRange = (
  currentTime: number,
  startTime: number,
  endTime: number
): boolean => {
  validateTime(currentTime)
  // Edge case: when endTime is less than startTime ex. 22:00-00:30
  if (startTime > endTime) {
    if (currentTime >= startTime && currentTime <= 1440) {
      return true
    } else if (currentTime < endTime) {
      return true
    } else return false
  }
  // normal case
  if (currentTime >= startTime && currentTime <= endTime) {
    return true
  } else return false
}
//--------------------------------------2
export const convertNumToTimeFormat = (
  a: number,
  b: 'HH' | 'mm'
): string | Error => {
  if (b === 'mm' && (a > 60 || a < 0)) {
    throw new Error('first Param must have value between 0-60')
  } else if (b === 'HH' && (a > 24 || a < 0)) {
    throw new Error('first Param must have value between 0-24')
  }

  return b === 'mm'
    ? dayjs().minute(a).format('mm')
    : b === 'HH'
    ? dayjs().hour(a).format('HH')
    : null
}
//--------------------------------------3
export const concatenateHrsMin = (time: Menu.Hour): string => {
  const { hrs, min } = time
  return `${convertNumToTimeFormat(hrs, 'HH')}:${convertNumToTimeFormat(
    min,
    'mm'
  )}`
}
/**
 *
 * @param currentTime client current time
 * @param targetTime is when update should happen: e.x startTime, endTime
 * @param interval determine intervals long of setInterval function based on how far currentTime away from targetTime
 */
//--------------------------------------4
export const getInterval = (
  currentTime: number,
  timeTargets: number[],
  interval?: number
): number => {
  const currentTargetTime = timeTargets.filter(
    (targetTime) =>
      targetTime - currentTime < 15 && targetTime - currentTime > 0
  )[0]
  if (!currentTargetTime) {
    return 60 * 60 // check every hour
  } else {
    return interval || 60 // check every minute
  }
}
//--------------------------------------5
export const withTimeRangeForDishes = (
  currentTime: number,
  startTime: number,
  endTime?: number
): boolean => {
  validateTime(currentTime)
  if (!endTime) {
    if (currentTime < startTime) {
      return false
    } else if (currentTime >= startTime) {
      return true
    }
  } else {
    if (currentTime < startTime || currentTime >= endTime) {
      return false
    } else if (currentTime >= startTime && currentTime <= endTime) {
      return true
    }
  }
}
