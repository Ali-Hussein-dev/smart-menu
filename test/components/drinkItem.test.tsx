import { render } from '../testUtils'
import MegaList from '../../database/index'
import DrinkItem from '../../src/components/drinks/DrinkItem'
import { Ctx } from '../../src/context'

const drinks = MegaList[0].menu.de.drinks
const softdrinkItem = drinks.drinks[0][1].list[0][1].itemsList[0]
const cocktailItem = drinks.cocktail[0][1].list[0][1].itemsList[0]
const longdrinkItem = drinks.drinks[6][1].list[1][1].itemsList[1]
const coffeeAddonItem = drinks.drinks[4][1].list[1][1].addon.addonList[1]

describe('DrinkItem', () => {
  test('softdrink item matches snapshot', () => {
    const { container } = render(
      <Ctx.Provider value={{ state: MegaList[0] }}>
        <DrinkItem item={softdrinkItem} sizeUnit={'L'} />
      </Ctx.Provider>
    )
    expect(container).toMatchSnapshot()
  })
  test('cocktail: non-hh time matches snapshot', () => {
    const { container } = render(
      <Ctx.Provider value={{ state: MegaList[0] }}>
        <DrinkItem item={cocktailItem} hasHH />
      </Ctx.Provider>
    )
    expect(container).toMatchSnapshot()
  })
  test('longdrink: matches snapshot', () => {
    const { container } = render(
      <Ctx.Provider value={{ state: MegaList[0] }}>
        <DrinkItem item={longdrinkItem} hasHH />
      </Ctx.Provider>
    )
    expect(container).toMatchSnapshot()
  })
  test('Coffee addon: matches snapshot', () => {
    const { container } = render(
      <Ctx.Provider value={{ state: MegaList[0] }}>
        <DrinkItem item={coffeeAddonItem} hasHH />
      </Ctx.Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
