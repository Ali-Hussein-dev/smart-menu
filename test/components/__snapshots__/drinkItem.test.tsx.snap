// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrinkItem Coffee addon: matches snapshot 1`] = `
<div>
  <div
    class="mb-3 text-blueGray-600"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="pr-6"
      >
        <span
          class="font-medium text-blueGray-700"
        >
          Sojamilch
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
      >
        <div
          class="flex items-center justify-center"
        >
          <span
            class="flex flex-col"
          >
            <span
              class="flex justify-center gap-x-1"
            >
              <span>
                0.50
              </span>
            </span>
          </span>
        </div>
        <div
          class="flex flex-col items-center w-full"
        >
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-1647zlw"
            role="separator"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DrinkItem cocktail: non-hh time matches snapshot 1`] = `
<div>
  <div
    class="mb-3 text-blueGray-600"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="pr-6"
      >
        <span
          class="font-medium text-blueGray-700"
        >
          Raspberry Gimlet
        </span>
        <div
          class="leading-5 break-words cursor-default text-blueGray-400"
          data-testid="text-expand"
        >
          <p
            class="chakra-text css-16ytai3"
          >
            Eigene Infusion aus Himbeere und Gin, mit Limette und Lime
          </p>
        </div>
      </div>
      <div
        class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
      >
        <div
          class="flex items-center justify-center"
        >
          <span
            class="flex flex-col"
          >
            <span
              class="flex justify-center gap-x-1"
            >
              <span>
                8.00
              </span>
            </span>
          </span>
        </div>
        <div
          class="flex flex-col items-center w-full"
        >
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-1647zlw"
            role="separator"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DrinkItem longdrink: matches snapshot 1`] = `
<div>
  <div
    class="mb-3 text-blueGray-600"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="pr-6"
      >
        <span
          class="font-medium text-blueGray-700"
        >
          Campari
        </span>
        <div
          class="leading-5 break-words cursor-default text-blueGray-400"
          data-testid="text-expand"
        >
          <p
            class="chakra-text css-16ytai3"
          >
            Maracuja | Soda | Tonic
          </p>
        </div>
      </div>
      <div
        class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
      >
        <div
          class="flex items-center justify-center"
        >
          <span
            class="flex flex-col"
          >
            <span
              class="flex justify-center gap-x-1"
            >
              <span>
                7.80
              </span>
            </span>
          </span>
        </div>
        <div
          class="flex flex-col items-center w-full"
        >
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-1647zlw"
            role="separator"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DrinkItem softdrink item matches snapshot 1`] = `
<div>
  <div
    class="mb-3 text-blueGray-600"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="pr-6"
      >
        <span
          class="font-medium text-blueGray-700"
        >
          Ananas
        </span>
      </div>
      <div
        class="flex flex-col items-center justify-between ml-1 text-blueGray-600"
      >
        <div
          class="flex items-center justify-center"
        >
          <span
            class="flex flex-col"
          >
            <span
              class="flex justify-center gap-x-1"
            >
              <span>
                3.50
              </span>
              <span
                class="text-blueGray-400"
              >
                0.3
                L
              </span>
            </span>
            <span
              class="flex justify-center gap-x-1"
            >
              <span>
                4.30
              </span>
              <span
                class="text-blueGray-400"
              >
                0.4
                L
              </span>
            </span>
          </span>
        </div>
        <div
          class="flex flex-col items-center w-full"
        >
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-1647zlw"
            role="separator"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
